### PR TITLE
feat(NcAppSidebar): Allow to set `open` state to prevent focus trap issues on mobile

### PR DESCRIPTION
**Pull request by susnux**
* Resolves https://github.com/nextcloud/forms/issues/2139
* Resolves https://github.com/nextcloud-libraries/nextcloud-vue/issues/5335

### ☑️ Resolves

If the current screen size is "mobile" (<= 512px) and the app sidebar is not removed from dom but only hidden (`v-show` instead of `v-if`) then currently a user can not interact with the page anymore.
This happens because a focus trap is registered but not removed when the element is hidden.

So this adds a comment about using `v-if` but for performance (e.g the sidebar is hidden and shown often) the sidebar also now allows to set the `open` prop for hiding it.

### 🏁 Checklist

- [ ] ⛑️ Tests are included or are not applicable
- [x] 📘 Component documentation has been extended, updated or is not applicable
- [ ] 3️⃣ Backport to `next` requested with a Vue 3 upgrade

### CODE REVIEW

**Comment by susnux** [comment]
BTW the styleguide still does not work because multiple sidebars are opened that always set a focus trap.
So the only solution to fix the styleguide I can currently think of is hiding all examples by default.

File: src/components/NcAppSidebar/NcAppSidebar.vue [hunk 1]
```diff
@@ -737,7 +823,7 @@ export default {
 		this.toggleFocusTrap()
 	},
 
-	beforeDestroy() {
+	beforeUnmount() {
```
Comment (ShGKme) [inline src/components/NcAppSidebar/NcAppSidebar.vue:826]: Have you checked this is working? I'm 99.99% sure new names are not backported to Vue 2.7...

**Comment by ShGKme** [comment]
> BTW the styleguide still does not work because multiple sidebars are opened that always set a focus trap. So the only solution to fix the styleguide I can currently think of is hiding all examples by default.

You mean, if  styleguidist is open on mobile?

**Comment by susnux** [comment]
> You mean, if styleguidist is open on mobile?

Yes, because currently one focus trap for each example is created and you can not interact with the page except for the first app sidebar example.

File: src/components/NcAppSidebar/NcAppSidebar.vue [hunk 2]
```diff
@@ -811,8 +897,14 @@ export default {
...
 			 *
 			 * @type {HTMLElement}
+			 * @deprecated Listen to `update:open` instead
 			 */
 			this.$emit('opening', element)
+			/**
+			 * Current open state emitted after the transitions are finished
+			 * @type {boolean}
+			 */
+			this.$emit('update:open', true)
```
[the remark this answers is not in the document]
Comment (susnux) [reply src/components/NcAppSidebar/NcAppSidebar.vue]: removed

File: src/components/NcAppSidebar/NcAppSidebar.vue [hunk 3]
```diff
@@ -827,8 +919,10 @@ export default {
 			 * The sidebar is closing and the transition is in progress
 			 *
 			 * @type {HTMLElement}
+			 * @deprecated Listen to `update:open` instead
 			 */
 			this.$emit('closing', element)
+			this.$emit('update:open', false)
```
[the remark this answers is not in the document]
Comment (susnux) [reply src/components/NcAppSidebar/NcAppSidebar.vue]: done

**Comment by susnux** [comment]
@nextcloud-libraries/designers about the toggle icon, we have this discussion here:
https://github.com/nextcloud-libraries/nextcloud-vue/issues/2804

So I used the icon recommended by @nimishavijay but got some feedback that it does not self descriptive for the sidebar.

Screenshot_20240513_161641

Maybe instead use `menu-open`? https://pictogrammers.com/library/mdi/icon/menu-open/

**Review by jancborchardt — CHANGES_REQUESTED** [review]
To make sure that it’s obvious and also different from the left navigation which uses menu-open, let’s use dock-right: https://pictogrammers.com/library/mdi/icon/dock-right/
dock-right

**Comment by susnux** [comment]
@jancborchardt implemented :heavy_check_mark:

File: src/components/NcAppSidebar/NcAppSidebar.vue [hunk 4]
```diff
@@ -366,155 +370,237 @@ A working alternative would be using an icon together with an `aria-label`:
...
 ```
-</docs>
 
+### Conditionally show the sidebar
+
+If the sidebar should be shown conditionally (e.g. using a button)
+and the users are expected to open and close the sidebar multiple times,
+then using `v-if` might result in bad performance.
+So instead use the `open` property:
+
```
Comment (ShGKme) [inline src/components/NcAppSidebar/NcAppSidebar.vue:380]: Can you also mention that when `open` prop is used, there is a built-in toggle button, you users don't have to implement it manually?

Comment (susnux) [reply src/components/NcAppSidebar/NcAppSidebar.vue:380]: Done :white_check_mark:

**Comment by ShGKme** [comment]
*Checking attrs inheritance*

File: src/components/NcAppSidebar/NcAppSidebar.vue [hunk 5]
```diff
@@ -366,155 +370,237 @@ A working alternative would be using an icon together with an `aria-label`:
...
+			:aria-label="t('Open sidebar')"
+			class="app-sidebar__toggle"
+			:class="toggleClasses"
+			type="tertiary"
+			@click="$emit('update:open', true)">
+			<template #icon>
+				<IconDockRight :size="20" />
+			</template>
+		</NcButton>
+		<transition appear
```
Comment (ShGKme) [inline src/components/NcAppSidebar/NcAppSidebar.vue:456]: This is a small breaking change
- Before `<aside>` through `<Transition>` was actually a root element receiving all passed attrs such as classes, styles or `data-cy`
- After it is a fragment, no attributes are passed.

So we need to bind `$attrs` and `$listeners`, and add `inheritAttrs: false`

**Review by ShGKme — CHANGES_REQUESTED** [review]

**Review by jancborchardt — APPROVED** [review]

**Review by ShGKme — APPROVED** [review]

**Pull request merged**